addons/payment_asiapay/models/payment_transaction.py

```diff
@@ -0,0 +1,139 @@
+            "payType": "N",
+            "lang": payment_asiapay_utils.get_lang(self.env.lang),
+            "payMethod": "ALL",
+        }
+        if self.acquirer_id.asiapay_currency_id.id != self.currency_id.id:
+            rendering_values.update({
+                "foreignCurrCode": CURRENCY_MAPPING.get(self.currency_id.name),
+            })
+        if self.acquirer_id.asiapay_secure_hash:
+            args = {
+                'hash_function': self.acquirer_id.asiapay_secure_hash_function,
+                'merchant_id': self.acquirer_id.asiapay_merchant_id,
+                'merchant_reference': order_ref,
+                'curr_code': CURRENCY_MAPPING.get(self.acquirer_id.asiapay_currency_id.name),
+                "amount": str(self.amount),
+                'payment_type': "N",
+                'secret': self.acquirer_id.asiapay_secure_hash,
+            }
+            secure_hash = payment_asiapay_utils.generate_secure_hash(**args)
+            rendering_values.update({
+                "secureHash": secure_hash,
+            })
+
+        rendering_values.update({
+            "api_url": self.acquirer_id._asiapay_get_api_url()
+        })
+        return rendering_values
+
+    @api.model
+    def _get_tx_from_feedback_data(self, provider, data):
+        tx = super()._get_tx_from_feedback_data(provider, data)
+        if provider != "asiapay":
+            return tx
+
+        reference = data.get('Ref')
+        if not reference:
+            raise ValidationError(
+                "AsiaPay: " + _(
+                    "Received data with missing reference %(r)s.",
+                    r=reference
+                )
+            )
+
+        order_reference = reference.split()[0]
+        tx = self.search([('reference', '=', order_reference), ('provider', '=', 'asiapay')])
+        if not tx:
+            raise ValidationError(
+                "AsiaPay: " + _("No transaction found matching reference %s.", order_reference)
+            )
+
+        return tx
+
+    def _process_feedback_data(self, data):
+        super()._process_feedback_data(data)
+        if self.provider != "asiapay":
+            return
+        if float_compare(float(data.get('Amt', '0.0')), self.currency_id._convert(self.amount, self.acquirer_id.asiapay_currency_id, self.company_id, fields.Date.today()), 2) != 0:
+            logging_values = {
+                'amount': data.get('Amt', '0.0'),
+                'total': self.currency_id._convert(self.amount, self.acquirer_id.asiapay_currency_id, self.company_id, fields.Date.today()),
+                'reference': self.reference,
+            }
+            _logger.error(
+                "the paid amount (%(amount)s) does not match the total + fees (%(total)s + "
+                "%(fees)s) for the transaction with reference %(reference)s", logging_values
+            )
+            raise ValidationError("AsiaPay: " + _("The amount does not match the total + fees."))
```

Comment:
Is this really necessary to check all of this? Both the payment and feedback data are signed, and the provider most likely does not allow updating the payment amount

Reply:
I was following Alipay when I did the development. I think we can skip this part